Subscribe customer to default payment plan after charging on stripe.

abannerjee opened the pull request:
A new customer is created using the Stripe API and the customer
is subscribed to the default payment plan defined in settings.

In addition, SHOP_STRIPE_API_VERSION is used to specify which
Stripe API version to use. Currently, djangoshop-stripe requires
a version higher than '2015-02-16' due to backward-incompatible
changes made on the charge object (charge.status returns 'succeeded'
instead of 'paid').

abannerjee committed:
Subscribe customer to default payment plan after charging on stripe.

A new customer is created using the Stripe API and the customer
is subscribed to the default payment plan defined in settings.

In addition, SHOP_STRIPE_API_VERSION is used to specify which
Stripe API version to use. Currently, djangoshop-stripe requires
a version higher than '2015-02-16' due to backward-incompatible
changes made on the charge object (charge.status returns 'succeeded'
instead of 'paid').

abannerjee commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Integrate Stripe-based payment methods into shop-core](https://trello.com/c/3BlYFt3m/1335-integrate-stripe-based-payment-methods-into-shop-core)

abannerjee committed:
Add model to store StripeCustomer

This model is used to link the site user to a
stripe customer created using the Stripe API.

abannerjee committed:
Add migrations for StripeCustomer model.

abannerjee committed:
Create a StripeCustomer and subscribe to plan after checkout.

After charging the StripePayment, get or create a StripeCustomer,
and subscribe StripeCustomer to the default subscription plan.

shaunbrady commented on shop_stripe/migrations/0001_initial.py:
```diff
@@ -0,0 +1,31 @@
+    initial = True
+
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
```
I was waiting to see this magic show up.  Nothing you did, but the AUTH_USER_MODEL has been a wrinkle for a while in Django-dom.....

shaunbrady commented on shop_stripe/payment.py:
```diff
@@ -52,6 +57,31 @@ def charge(self, cart, request):
+        try:
+            customer = StripeCustomer.objects.get(user_id=user_id)
+            stripe_customer = stripe.Customer.retrieve(customer.stripe_customer_id)
+        except ObjectDoesNotExist:
```
One suggestion here. This code is fine, but I would probably not use the generic ObjectDoesNotExist.  Each Django ORM model (StripeCustomer in this case? (This ambiguity will prove my point in a secont)) has it's own .DoesNotExist (double check I have the exception name right).  If you only expect to ever see StripeCustomer not existing, use it directly as in `except StripeCustomer.DoesNotExist`.  The reason is, you are more clear about what you're doing.

One other small issue, and I'm of two minds of this:  Some frown on using try/except as control blocks (if/else).  The code here reads fine, and actually is efficient in the sense of "ask for forgiveness instead of permission".  @amadornimbis can you weigh in?

Does stripe.Customer have something like a get_or_create()?  I think the Django ORM does, so you might have an opportunity there to streamline things.

I've tossed about 2.5 things up there, that are mostly only spatially related.  Let me know if I can be more clear.

shaunbrady replied:
... honestly I've used try/except in this way.... so I'm probably calling the kettle black at best....  but I shouldn't be your bar for quality.

abannerjee replied:
Thanks for the feedback! I agree with you about not using the generic ObjectDoesNotExist, and instead I'll change it to expect StripeCustomer.DoesNotExist instead.

Regarding using try/except as control blocks. The original code had the same approach so I kept it, and I believe it is because it seems stripe.Customer doesn't have a get_or_create method. While testing, I noticed I was able to create the same customer twice given identical parameters.

In addition, stripe seems to be very limited when it comes to retrieving customers as well (can only query on stripe customer id).
Stripe API for reference: https://stripe.com/docs/api/python#create_customer

shaunbrady commented on shop_stripe/payment.py:
```diff
@@ -22,8 +24,12 @@ def get_payment_request(self, cart, request):
         From the given request, add a snippet to the page.
         """
+        stripe.api_key = settings.SHOP_STRIPE['APIKEY']
+        stripe.api_version = settings.SHOP_STRIPE_API_VERSION
```
Did we add this setting?  Should we put it in the SHOP_STRIPE dictionary?  (SHOP_STRIP['API_VERSION'])  I could be convinced against this if the original authors ever (or might?) iterate the dictionary.  Debate internally, happy with either decision.

abannerjee replied:
This is a new setting I am adding to SHOP_STRIPE. I was debating on whether it should go into SHOP_STRIPE or not, and played it safe by having it separate.

Since it seems we would want to push these changes upstream eventually, and because it looks like adding that setting into SHOP_STRIPE won't break anything (at least for now), I'll make the change and add it to that dictionary.

ericamador commented on shop_stripe/models.py:
```diff
@@ -0,0 +1,28 @@
+	"""
+
+	class Meta:
+		verbose_name = "Stripe Customer"
```
It looks like tabs are being used in this file. These should be replaced with four spaces.

abannerjee replied:
Good catch, thanks! This should be fixed now.

abannerjee committed:
Expect StripeCustomer.DoesNotExist instead of ObjectDoesNotExist

Makes the code more clear by stating we may expect StripeCustomer
to not exist.

abannerjee committed:
Move setting 'API_VERSION' into SHOP_STRIPE dictionary.

Previously, SHOP_STRIPE_API_VERSION existed as its own
setting. Since adding it to SHOP_STRIPE will not break
existing code, it would make more sense to group all
SHOP_STRIPE settings together.
